### ISSUE TITLE
loss_scale variance across DP ranks **may** introduce slight gradient bias in RL training

### ISSUE THREAD

**Issue by ijamil1** [PrimeIntellect-ai/prime-rl#2358]
## Relevant File

src/prime_rl/trainer/rl/train.py

## Observation

When running RL training with data parallelism (DP), each DP rank computes a local `loss_scale` — the number of loss-contributing tokens on that rank for the current step:

```python
loss_scale = sum(micro_batch["loss_mask"].sum().item() for micro_batch in micro_batches)
loss_scale = max(loss_scale, 1)
```

This value is passed to `compute_loss`, which divides the summed token loss by it before `.backward()`:

```python
scaled_loss = total_loss / loss_scale
```

Because different DP ranks process different samples with different lengths, `loss_scale` can vary across ranks. I verified this empirically by all-gathering `loss_scale` across 2 `dp` ranks and tracking `loss_scale/max_ratio` in W&B, where `loss_scale/max_ratio` is simply the maximum local `loss_scale` divided by the minimum local `loss_scale` for that step (across the 2 DP ranks). We observed occasional large values, e.g. ~1.02 and ~1.04, indicating step-level skew between ranks.

**Experimental setup:** the Hendrycks Sanity environment (`examples/hendrycks_sanity/rl.toml`) with the following modifications: `max_steps=50`, `num_train_gpus=2`, `num_infer_gpus=1`, `batch_size=200`, `trainer.model.seq_len=8192`, and activation checkpointing enabled every 2 layers. Including link to the loss_scale/max_ratio plot below:

![loss_scale:max_ratio plot](https://github.com/ijamil1/prime-rl/blob/loss-scale-variance-diagnostic/loss_scale%3Amax_ratio.png)

## Why this may be a problem (not completely sure!)

Without loss of generality, assume CP=1. When loss.backward() is called after the microbatch loop and because of the FSDP gradient scatter-reduce, the effective gradient is:

```
param.grad = (1 / num dp ranks) * sum_r [ (1 / loss_scale_r) * sum_m g_{r,m} ]
```

where `g_{r,m}` is the per-token gradient sum on rank `r`, microbatch `m`, and `loss_scale_r` is pulled outside the microbatch sum because it is a step-level constant on each rank. I believe this is not a simple per-token mean across all tokens in the batch when loss_scale_r is not uniform across the DP ranks. Ranks with fewer loss tokens (smaller `loss_scale_r`) contribute a larger weight per token, upweighting their gradients relative to ranks with more tokens.

I believe the unbiased per-token mean would be:

```
param.grad = sum_r sum_m g_{r,m} / sum_r loss_scale_r
```

## A possible fix

All-reduce `loss_scale` across DP ranks before the microbatch loop to get the global token count contributing to the loss per step, pass that to `compute_loss` as the denominator, and multiply gradients by `dp_world_size` after backward to undo FSDP's divide:

```python
global_token_count = all_reduce(loss_scale, op=SUM)  # before microbatch loop
loss = compute_loss(..., loss_scale=global_token_count)
loss.backward()
# after microbatch loop:
param.grad *= dp_world_size
```

This should ensure every token contributes equally to the gradient regardless of which DP rank it was assigned to.

## Proposed fix (PR forthcoming)

I have implemented this fix in a branch and will open a PR linked to this issue shortly. The implementation:

1. All-reduces `loss_scale` across `dp` ranks before the microbatch loop to obtain the global token count.
2. Passes the global token count to `compute_loss` as the normalization denominator (replacing the local `loss_scale`).
3. After the microbatch loop, multiplies each parameter's gradient by `dp_world_size` to undo FSDP's built-in division and recover the true per-token mean gradient.

The change is minimal and confined to `src/prime_rl/trainer/rl/train.py`.

## Note
I am not sure I am correct. Even if I am correct, the loss_scale skew is random over tokens/sequences and so it may wash out/not be consequential. Would love for feedback!

**Comment by mikasenghaas** [comment]
hi @ijamil1, thx for the detailed report and sorry for only getting back at you so late! indeed, we do a mean of mean which may not be the same as a global mean. we did this for ease of impementaiton initially. the multiplication by `dp_world_size` is a neat trick. if you can show that loss scale is in the same ballpark, as before, this should be a safe merge from my side. wdyt @faresobeid @samsja ?

**Comment by ijamil1** [comment]
@mikasenghaas Happy to see the issue is closed! sorry for getting back to this a bit late. Between me opening this issue and you guys working on it, I had also done some tests to show the fix was valid. **Given the change from @samsja was merged into main, my results are not needed at this point. But, since I had already done some experiments a week or two ago, I thought I’d share**:


I did some follow-up diagnostics to validate that the loss-scale normalization fix moves distributed gradients closer to the 1-GPU ground-truth reference.

Setup:

- Reference / ground truth: `(DP=1, CP=1)` RL run for 20 steps on the Hendrycks Sanity env. Since this uses a single GPU, there is no cross-rank local `loss_scale` skew. At each step, I saved the pre-step LoRA weights, exact prompt/completion microbatches, advantages, logprobs, and LoRA gradients.
- Comparisons: I compared the delta between:
  - gradients / optimizer updates computed with the fix I had proposed vs. those of the reference, and
  - gradients / optimizer updates computed with the existing local-`loss_scale` implementation vs. those of the reference.
- Distributed replay setups:
  - `(DP=2, CP=1)`
  - `(DP=3, CP=1)`
  - `(DP=2, CP=2)`

The replay setup was designed to isolate the loss-scale bug and avoid confounders from rollout sampling, model drift, or batch differences. For each replay step, both the fixed and existing implementations load the same reference pre-step model weights, load the exact same saved batches/microbatches/logprobs/advantages, and route the data to the corresponding DP ranks in the same way. No new rollouts are sampled and model weights are not advanced across steps. So the remaining differences should come from the gradient computation method itself, plus expected floating-point drift from multi-GPU execution.

The results JSON files are available here: [ijamil1/prime-rl `gradient_comparisons`](https://github.com/ijamil1/prime-rl/tree/chore/grad-replay-main/gradient_comparisons).

Results:

- For `(DP=2, CP=1)` and `(DP=3, CP=1)`, the replay diagnostics consistently show that the fix is closer to the `(DP=1, CP=1)` reference than the existing local-`loss_scale` implementation.
- This holds for both raw LoRA gradient comparisons and the AdamW update deltas implied by those gradients.
- Across the replayed steps, the fix is better across cosine similarity, relative L2 error, and per-parameter/per-element distance comparisons.
- In other words, when only DP is enabled, the fix moves the distributed gradient computation noticeably closer to the single-GPU ground truth, and that improvement also survives the optimizer-update calculation.
- For `(DP=2, CP=2)`, the result is more modest. The fix is slightly closer to the reference overall, but the improvement is much smaller than in the DP-only cases.

Overall, these replay diagnostics strongly support the local-`loss_scale` skew issue and the normalization fix (which I’m sure you guys have already determined!) **Again, just sharing for transparency and not raising anything new.** Thanks!

**Comment by mikasenghaas** [comment]
doing lords work! thx for bringing this to our attention again. neat fix for sure!
